Better debug logs for borrowck constraint graph

b-naber opened the pull request:
It's really cumbersome to work with `RegionVar`s when trying to debug borrowck code or when trying to understand how the borrowchecker works. This PR collects some region information (behind `cfg(debug_assertions)`) for created `RegionVar`s (NLL region vars, this PR doesn't touch canonicalization) and prints the nodes and edges of the strongly connected constraints graph using representatives that use that region information (either lifetime names, locations in MIR or spans).

b-naber committed:
add BorrowckInferCtxt

b-naber committed:
collect region contexts during mir renumbering

b-naber committed:
collect existentials and placeholders

b-naber committed:
sccs info

b-naber committed:
some conditional imports

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-13`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/3436293824/jobs/5729651884) [(plain)](https://github.com/rust-lang/rust/commit/fabea6e2a975ebdffcbcf2911f075ce36f6ca534/checks/9410274927/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
   Compiling rustc_hir_analysis v0.0.0 (/checkout/compiler/rustc_hir_analysis)
   Compiling rustc_codegen_ssa v0.0.0 (/checkout/compiler/rustc_codegen_ssa)
   Compiling rustc_traits v0.0.0 (/checkout/compiler/rustc_traits)
   Compiling rustc_borrowck v0.0.0 (/checkout/compiler/rustc_borrowck)
error[E0407]: method `replace_late_bound_regions_with_nll_infer_vars` is not a member of trait `InferCtxtExt`
    |
    |
894 |       #[instrument(skip(self, indices))]
    |       |
    |       not a member of trait `InferCtxtExt`
    |       in this procedural macro expansion
    |       in this procedural macro expansion
895 |       fn replace_late_bound_regions_with_nll_infer_vars(
    |          ---------------------------------------------- help: there is an associated function with a similar name: `replace_bound_regions_with_nll_infer_vars`
   ::: /cargo/registry/src/github.com-1ecc6299db9ec823/tracing-attributes-0.1.22/src/lib.rs:552:1
    |
552 | / pub fn instrument(
553 | |     args: proc_macro::TokenStream,
553 | |     args: proc_macro::TokenStream,
554 | |     item: proc_macro::TokenStream,
555 | | ) -> proc_macro::TokenStream {
    | |____________________________- in this expansion of `#[instrument]`

error[E0425]: cannot find function `for_each_late_bound_region_defined_on` in this scope
    |
    |
901 |           for_each_late_bound_region_defined_on(self.tcx, typeck_root_def_id, |r| {
    |           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: a function with a similar name exists: `for_each_late_bound_region_in_item`
...
982 | / fn for_each_late_bound_region_in_item<'tcx>(
983 | |     tcx: TyCtxt<'tcx>,
984 | |     mir_def_id: LocalDefId,
985 | |     mut f: impl FnMut(ty::Region<'tcx>),
996 | |     }
997 | | }
997 | | }
    | |_- similarly named function `for_each_late_bound_region_in_item` defined here

error[E0046]: not all trait items implemented, missing: `replace_late_bound_regions_with_nll_infer_vars_in_recursive_scope`
    |
    |
734 | /     fn replace_late_bound_regions_with_nll_infer_vars_in_recursive_scope(
735 | |         &self,
736 | |         mir_def_id: LocalDefId,
737 | |         indices: &mut UniversalRegionIndices<'tcx>,
738 | |     );
    | |______- `replace_late_bound_regions_with_nll_infer_vars_in_recursive_scope` from trait
...
747 |   impl<'cx, 'tcx> InferCtxtExt<'tcx> for BorrowckInferCtxt<'cx, 'tcx> {
    |   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ missing `replace_late_bound_regions_with_nll_infer_vars_in_recursive_scope` in implementation
error[E0023]: this pattern has 1 field, but the corresponding tuple variant has 2 fields
   --> compiler/rustc_borrowck/src/type_check/relate_tys.rs:135:45
    |
    |
135 |                 ty::BoundRegionKind::BrAnon(_) => Symbol::intern("anon"),
    |
   ::: /checkout/compiler/rustc_middle/src/ty/sty.rs:62:12
    |
    |
62  |     BrAnon(u32, Option<Span>),
    |
help: use `_` to explicitly ignore each field
    |
    |
135 |                 ty::BoundRegionKind::BrAnon(_, _) => Symbol::intern("anon"),
help: use `..` to ignore all fields
    |
    |
135 |                 ty::BoundRegionKind::BrAnon(..) => Symbol::intern("anon"),

error[E0023]: this pattern has 1 field, but the corresponding tuple variant has 2 fields
    --> compiler/rustc_borrowck/src/type_check/mod.rs:1431:57
     |
     |
1431 | ...                   ty::BoundRegionKind::BrAnon(_) => Symbol::intern("anon"),
     |
    ::: /checkout/compiler/rustc_middle/src/ty/sty.rs:62:12
     |
     |
62   |     BrAnon(u32, Option<Span>),
     |
help: use `_` to explicitly ignore each field
     |
     |
1431 |                             ty::BoundRegionKind::BrAnon(_, _) => Symbol::intern("anon"),
help: use `..` to ignore all fields
     |
     |
1431 |                             ty::BoundRegionKind::BrAnon(..) => Symbol::intern("anon"),

error[E0061]: this function takes 2 arguments but 1 argument was supplied
   --> compiler/rustc_borrowck/src/universal_regions.rs:446:53
    |
    |
446 |                         let region_vid = self.infcx.next_nll_region_var(FR);
    |                                                     ^^^^^^^^^^^^^^^^^^^---- an argument of type `RegionCtxt` is missing
note: associated function defined here
   --> compiler/rustc_borrowck/src/lib.rs:546:19
    |
    |
546 |     pub(crate) fn next_nll_region_var(
547 |         &self,
547 |         &self,
548 |         origin: NllRegionVariableOrigin,
    |         -------------------------------
549 |         ctxt: RegionCtxt,
help: provide the argument
    |
    |
446 |                         let region_vid = self.infcx.next_nll_region_var(FR, /* RegionCtxt */);

error[E0061]: this function takes 2 arguments but 1 argument was supplied
   --> compiler/rustc_borrowck/src/universal_regions.rs:474:45
    |
    |
474 |                 let region_vid = self.infcx.next_nll_region_var(FR);
    |                                             ^^^^^^^^^^^^^^^^^^^---- an argument of type `RegionCtxt` is missing
note: associated function defined here
   --> compiler/rustc_borrowck/src/lib.rs:546:19
    |
    |
546 |     pub(crate) fn next_nll_region_var(
547 |         &self,
547 |         &self,
548 |         origin: NllRegionVariableOrigin,
    |         -------------------------------
549 |         ctxt: RegionCtxt,
help: provide the argument
    |
    |
474 |                 let region_vid = self.infcx.next_nll_region_var(FR, /* RegionCtxt */);

error[E0061]: this function takes 2 arguments but 1 argument was supplied
   --> compiler/rustc_borrowck/src/universal_regions.rs:877:39
    |
    |
877 |                 let region_vid = self.next_nll_region_var(FR);
    |                                       ^^^^^^^^^^^^^^^^^^^---- an argument of type `RegionCtxt` is missing
note: associated function defined here
   --> compiler/rustc_borrowck/src/lib.rs:546:19
    |
    |
546 |     pub(crate) fn next_nll_region_var(
547 |         &self,
547 |         &self,
548 |         origin: NllRegionVariableOrigin,
    |         -------------------------------
549 |         ctxt: RegionCtxt,
help: provide the argument
    |
    |
877 |                 let region_vid = self.next_nll_region_var(FR, /* RegionCtxt */);

Some errors have detailed explanations: E0023, E0046, E0061, E0407, E0425.
For more information about an error, try `rustc --explain E0023`.
error: could not compile `rustc_borrowck` due to 8 previous errors
```

</details>

Nilstrieb commented:
This adds a huge amount of `#[cfg]`ed code, which isn't very good. I think most of this can be replaced with `cfg!(debug_assertions)` or other control-flow based conditional execution instead, which is way easier to maintain.
With something like
```rust
fn with_reg_var_to_origin(&self, f: impl FnOnce(RefCell<FxHashMap<ty::RegionVid, RegionCtxt>>))
```
that calls the closure when `debug_assertions` are enabled and is a noop when they aren't, the `#[cfg]` attributes could be contained to a minimum (one for the field, one for `new` and one for something like the function above).

b-naber committed:
rebase

bors commented:
:umbrella: The latest upstream changes (presumably #104293) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

jackh726 commented:
Can we try to add this without all the `cfg`s? And measure perf? Then add as few `cfg`s as possible to not regress perf?

b-naber committed:
remove cfgs

b-naber commented:
> Can we try to add this without all the `cfg`s? And measure perf? Then add as few `cfg`s as possible to not regress perf?

@jackh726 Sorry, didn't immediately have time when I saw your reply and then forgot that you posted.

bors commented:
:umbrella: The latest upstream changes (presumably #105036) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

jackh726 commented:
Agh sorry for the delay. Can you rebase and remove the cargo.lock changes?

albertlarsan68 commented:
@rustbot label -A-bootstrap

jackh726 commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

jackh726 commented:
Agh

jackh726 commented:
@bors try

bors commented:
:hourglass: Trying commit 570ad62318962df7f7a39b0def98496c75d1995c with merge ac1fb8505edc6653abdc5e92f701a76fbe64e0ea...
<!-- homu: {"type":"TryBuildStarted","head_sha":"570ad62318962df7f7a39b0def98496c75d1995c","merge_sha":"ac1fb8505edc6653abdc5e92f701a76fbe64e0ea"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/3950477026/jobs/6765506836)
Build commit: ac1fb8505edc6653abdc5e92f701a76fbe64e0ea (`ac1fb8505edc6653abdc5e92f701a76fbe64e0ea`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/3950477026/jobs/6765506836"},"merge_sha":"ac1fb8505edc6653abdc5e92f701a76fbe64e0ea"} -->

rust-timer commented:
Queued ac1fb8505edc6653abdc5e92f701a76fbe64e0ea with parent 1f72129ffe5e8c495113f9a2d4e1730f7fad3209, future [comparison URL](https://perf.rust-lang.org/compare.html?start=1f72129ffe5e8c495113f9a2d4e1730f7fad3209&end=ac1fb8505edc6653abdc5e92f701a76fbe64e0ea).
<!-- rust-timer: temporary -->

rust-timer commented:
Finished benchmarking commit (ac1fb8505edc6653abdc5e92f701a76fbe64e0ea): [comparison URL](https://perf.rust-lang.org/compare.html?start=1f72129ffe5e8c495113f9a2d4e1730f7fad3209&end=ac1fb8505edc6653abdc5e92f701a76fbe64e0ea&stat=instructions:u).

### Overall result: ❌ regressions - ACTION NEEDED


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

**Next Steps**: If you can justify the regressions found in this try perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please fix the regressions and do another perf run. If the next run shows neutral or positive results, the label will be automatically removed.

@bors rollup=never
@rustbot label: -S-waiting-on-perf +perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=1f72129ffe5e8c495113f9a2d4e1730f7fad3209&end=ac1fb8505edc6653abdc5e92f701a76fbe64e0ea&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 1.1% | [0.2%, 3.0%] | 57    |
| Regressions ❌ <br /> (secondary)  | 1.1% | [0.5%, 2.1%] | 30    |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 1.1% | [0.2%, 3.0%] | 57    |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=1f72129ffe5e8c495113f9a2d4e1730f7fad3209&end=ac1fb8505edc6653abdc5e92f701a76fbe64e0ea&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.6%  | [0.6%, 0.6%]   | 1     |
| Regressions ❌ <br /> (secondary)  | 2.1%  | [1.4%, 2.7%]   | 7     |
| Improvements ✅ <br /> (primary)   | -5.7% | [-5.7%, -5.7%] | 1     |
| Improvements ✅ <br /> (secondary) | -2.4% | [-3.1%, -1.3%] | 5     |
| All ❌✅ (primary)                 | -2.6% | [-5.7%, 0.6%]  | 2     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=1f72129ffe5e8c495113f9a2d4e1730f7fad3209&end=ac1fb8505edc6653abdc5e92f701a76fbe64e0ea&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 1.7% | [0.7%, 3.8%] | 27    |
| Regressions ❌ <br /> (secondary)  | 2.3% | [1.4%, 4.7%] | 17    |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 1.7% | [0.7%, 3.8%] | 27    |
</details>

jackh726 reviewed (changes requested):
Okay, so perf isn't that bad, but probably worth doing a few small things to reduce the impact in non-debug mode.

First thing I would do is split out the trivial changes into a separate PR. This would include debug deriving, adding instrument calls, and maybe the addition of BorrowckInferCtxt as a thin wrapper around InferCtxt.

After that, one easy thing to do would be to have next_nll_region_var take a function returning RegionCtxt, which only gets called in debug mode.

jackh726 commented:
@b-naber are you still interested in this PR? I could potentially take it over if you're unable to continue working on it.

b-naber committed:
add some cfgs back

b-naber commented:
@jackh726 Sorry for completely neglecting this PR, I'll try to finish this.

Before splitting this into other PRs, can we run perf again (I've added back some cfg statements and changed `next_nll_region_var` to take a function as you suggested)?

jackh726 commented:
No worries.

Yes of course.

@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

bors commented:
:hourglass: Trying commit 0725d0ceee13d3484b2b7117e4cff82c01d8f868 with merge b881ed0cec07b0a564e87aba198a3bb93516356f...
<!-- homu: {"type":"TryBuildStarted","head_sha":"0725d0ceee13d3484b2b7117e4cff82c01d8f868","merge_sha":"b881ed0cec07b0a564e87aba198a3bb93516356f"} -->

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/4218447517/jobs/7322964788) [(plain)](https://github.com/rust-lang/rust/commit/0725d0ceee13d3484b2b7117e4cff82c01d8f868/checks/11452723875/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
    Checking rustc_mir_transform v0.0.0 (/checkout/compiler/rustc_mir_transform)
error[E0423]: expected function, found macro `format`
   --> compiler/rustc_borrowck/src/region_infer/mod.rs:277:30
    |
277 |         components_str.push(&format(
    |
help: use `!` to invoke the macro
    |
    |
277 |         components_str.push(&format!(
help: consider importing one of these items instead
    |
1   | use rustc_span::sym::format;
    |
    |
1   | use std::fmt::format;
    |

error[E0599]: no method named `push` found for reference `&str` in the current scope
    |
    |
277 |         components_str.push(&format(
    |                        ^^^^ method not found in `&str`
Some errors have detailed explanations: E0423, E0599.
For more information about an error, try `rustc --explain E0423`.
error: could not compile `rustc_borrowck` due to 2 previous errors
warning: build failed, waiting for other jobs to finish...
```

</details>

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/4218496448/jobs/7324181867)
Build commit: b881ed0cec07b0a564e87aba198a3bb93516356f (`b881ed0cec07b0a564e87aba198a3bb93516356f`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/4218496448/jobs/7324181867"},"merge_sha":"b881ed0cec07b0a564e87aba198a3bb93516356f"} -->

rust-timer commented:
Queued b881ed0cec07b0a564e87aba198a3bb93516356f with parent 7aa413d59206fd511137728df3d9e0fd377429bd, future [comparison URL](https://perf.rust-lang.org/compare.html?start=7aa413d59206fd511137728df3d9e0fd377429bd&end=b881ed0cec07b0a564e87aba198a3bb93516356f).
<!-- rust-timer: temporary -->

rust-timer commented:
Finished benchmarking commit (b881ed0cec07b0a564e87aba198a3bb93516356f): [comparison URL](https://perf.rust-lang.org/compare.html?start=7aa413d59206fd511137728df3d9e0fd377429bd&end=b881ed0cec07b0a564e87aba198a3bb93516356f&stat=instructions:u).

### Overall result: no relevant changes - no action needed


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

@bors rollup=never
@rustbot label: -S-waiting-on-perf -perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=7aa413d59206fd511137728df3d9e0fd377429bd&end=b881ed0cec07b0a564e87aba198a3bb93516356f&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=7aa413d59206fd511137728df3d9e0fd377429bd&end=b881ed0cec07b0a564e87aba198a3bb93516356f&stat=max-rss)
This benchmark run did not return any relevant results for this metric.

### [Cycles](https://perf.rust-lang.org/compare.html?start=7aa413d59206fd511137728df3d9e0fd377429bd&end=b881ed0cec07b0a564e87aba198a3bb93516356f&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.

jackh726 commented:
Okay good. Can we do one more with the cfgs removed?

b-naber committed:
remove cfg attributes

b-naber commented:
Forgot that I also should be able to request a perf run, let's see:

@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

bors commented:
:hourglass: Trying commit c9843d61440edb19642920f2f8f33652e9c83971 with merge e02ff57926ffbe7e5ed3694d530b665c6917be1d...
<!-- homu: {"type":"TryBuildStarted","head_sha":"c9843d61440edb19642920f2f8f33652e9c83971","merge_sha":"e02ff57926ffbe7e5ed3694d530b665c6917be1d"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/4227348455/jobs/7343035968)
Build commit: e02ff57926ffbe7e5ed3694d530b665c6917be1d (`e02ff57926ffbe7e5ed3694d530b665c6917be1d`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/4227348455/jobs/7343035968"},"merge_sha":"e02ff57926ffbe7e5ed3694d530b665c6917be1d"} -->

rust-timer commented:
Queued e02ff57926ffbe7e5ed3694d530b665c6917be1d with parent 5243ea5c29b136137c36bd773e5baa663790e097, future [comparison URL](https://perf.rust-lang.org/compare.html?start=5243ea5c29b136137c36bd773e5baa663790e097&end=e02ff57926ffbe7e5ed3694d530b665c6917be1d).
<!-- rust-timer: temporary -->

rust-timer commented:
Finished benchmarking commit (e02ff57926ffbe7e5ed3694d530b665c6917be1d): [comparison URL](https://perf.rust-lang.org/compare.html?start=5243ea5c29b136137c36bd773e5baa663790e097&end=e02ff57926ffbe7e5ed3694d530b665c6917be1d&stat=instructions:u).

### Overall result: ❌ regressions - no action needed


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

@bors rollup=never
@rustbot label: -S-waiting-on-perf -perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=5243ea5c29b136137c36bd773e5baa663790e097&end=e02ff57926ffbe7e5ed3694d530b665c6917be1d&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -    | -            | 0     |
| Regressions ❌ <br /> (secondary)  | 0.2% | [0.2%, 0.2%] | 1     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | -    | -            | 0     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=5243ea5c29b136137c36bd773e5baa663790e097&end=e02ff57926ffbe7e5ed3694d530b665c6917be1d&stat=max-rss)
This benchmark run did not return any relevant results for this metric.

### [Cycles](https://perf.rust-lang.org/compare.html?start=5243ea5c29b136137c36bd773e5baa663790e097&end=e02ff57926ffbe7e5ed3694d530b665c6917be1d&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -    | -            | 0     |
| Regressions ❌ <br /> (secondary)  | 2.6% | [2.2%, 3.1%] | 2     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | -    | -            | 0     |
</details>

jackh726 commented:
Okay, unless there's anything else you think you want to try, I think we land this as-is. How's that sound @b-naber?

b-naber commented:
@jackh726 Don't have anything else to try.

jackh726 commented on compiler/rustc_borrowck/src/renumber.rs:
```diff
@@ -29,27 +31,68 @@ pub fn renumber_mir<'tcx>(
     infcx.tcx.fold_regions(value, |_region, _depth| {
         let origin = NllRegionVariableOrigin::Existential { from_forall: false };
-        infcx.next_nll_region_var(origin)
+        infcx.next_nll_region_var(origin, || get_ctxt_fn())
```
Can just pass `get_ctxt_fn` right?

b-naber replied:
No, `get_ctxt_fn` is captured in `fold_regions` which is `FnMut`, so we can't move there.

jackh726 reviewed (approved):
Just a couple nits. r=me with or without them

b-naber committed:
address review

b-naber commented:
@bors r=jackh726 rollup
